
# Data Cleaning for phase two clinical trial ( Auralin vas Novodra )
 

- This dataset was constructed with the consult of real doctors to ensure plausibility.
-This clinical trial data for an alternative insulin was inspired and closely mimics this real clinical trial for an inhaled insulin called Afrezza.
- The data quality issues in this dataset mimic real, common data quality issues in healthcare data. These issues impact quality of care, patient registration, and revenue.
- The patients in this dataset were created using this fake name generator and do not include real names, addresses, phone numbers, emails, etc.


## Gather

In [29]:
import pandas as pd

In [30]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [31]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [32]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [33]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [34]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 37.4+ KB


In [35]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 11.0+ KB


In [36]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 472.0+ bytes


In [38]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [39]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [40]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [41]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [42]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [43]:
patients.sample(5)

patient_id assigned_sex  given_name    surname                 address  \
179         180         male     Dominik  Grunewald     3574 Park Boulevard   
66           67         male  Konstantyn  Kowalczyk     4707 Parkway Street   
230         231       female   Elisabeth    Knudsen                     NaN   
463         464       female       Bouke     Glaser        3006 Maple Court   
410         411         male      Nathan   Cumpston  1203 Benson Park Drive   

             city       state  zip_code        country  \
179  Marshalltown          IA   50158.0  United States   
66      San Diego  California   92103.0  United States   
230           NaN         NaN       NaN            NaN   
463    Owensville          MO   65066.0  United States   
410         Wayne          OK   73095.0  United States   

                                         contact   birthdate  weight  height  \
179        641-753-5678DominikGrunewald@cuvox.de  11/20/1935   143.0      71   
66   760-287-8295KonstantynKowalczyk@armyspy.com    3/2/1938   194.3      67   
230                                          NaN   9/23/1976   165.9      63   
463         BoukeGlaser@einrot.com1 573 437 7334  10/21/1996   142.1      66   
410       NathanCumpston@rhyta.com1 405 449 7960   10/6/1965   178.0      67   

      bmi  
179  19.9  
66   30.4  
230  29.4  
463  22.9  
410  27.9

In [44]:
patients.surname.value_counts()

Doe         6
Jakobsen    3
Taylor      3
Nilsen      2
Kadyrov     2
           ..
Glockner    1
Bazarova    1
Mehari      1
Bois        1
Morita      1
Name: surname, Length: 466, dtype: int64

In [45]:
patients.address.value_counts()

123 Main Street           6
2476 Fulton Street        2
2778 North Avenue         2
648 Old Dear Lane         2
664 Lyon Avenue           1
                         ..
4040 Linda Street         1
1128 Jacobs Street        1
4649 Worley Avenue        1
3008 Walkers Ridge Way    1
1094 Jones Avenue         1
Name: address, Length: 483, dtype: int64

In [46]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [47]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [48]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [49]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [50]:
sum(treatments.auralin.isnull())

0

In [51]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [52]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

#### `treatments`: Missing records (280 instead of 350)

##### Define
*Your definition here. Note: the missing `treatments` records are stored in a file named `treatments_cut.csv`, which you can see in this Jupyter Notebook's dashboard (click the **jupyter** logo in the top lefthand corner of this Notebook). Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) for the function used in the solution.*

##### Code

In [53]:
# Your cleaning code here
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_clean = pd.concat([treatments_clean,treatments_cut],ignore_index=True)

##### Test

In [54]:
# Your testing code here
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [55]:
treatments_clean.tail()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
*Recalculate the `hba1c_change` column: `hba1c_start` minus `hba1c_end`. *

##### Code

In [56]:
# Your cleaning code here
treatments_clean.hba1c_change = (treatments_clean.hba1c_start - 
                                 treatments_clean.hba1c_end)

##### Test

In [57]:
# Your testing code here
treatments_clean.hba1c_change.head()

0    0.43
1    0.47
2    0.43
3    0.35
4    0.32
Name: hba1c_change, dtype: float64

### Tidiness

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
*Your definition here. Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). Hint 3: [email address regex pattern](http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

##### Code

In [58]:
# Your cleaning code here
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

# [a-zA-Z] to signify emails in this dataset all start and end with letters
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)


##### Test

In [59]:
# Your testing code here
# Confirm contact column is gone
list(patients_clean)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone_number',
 'email']

In [60]:
patients_clean.phone_number.sample(25)

40          404 547 4508
480         210 326 5509
344         404-713-3641
29     +1 (845) 858-7707
380         609-641-2070
267    +1 (785) 823-6728
158    +1 (918) 971-5864
352         912-988-6655
60          619-570-3898
103    +1 (407) 838-0201
226         914-784-6646
468         573-728-3861
0           951-719-9170
414         937-526-1152
67          508-804-4850
48          312-719-7238
20          781 447 1763
213         917-622-9142
485         850 659 0417
406    +1 (706) 581-5226
69     +1 (773) 615-9328
458         775-848-6970
273         601-848-3093
331         323 635 9919
483         731-577-0292
Name: phone_number, dtype: object

In [61]:
# Confirm that no emails start with an integer (regex didn't match for this)
patients_clean.email.sort_values().head()

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
Name: email, dtype: object

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
*Your definition here. Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

**Melt the auralin and novodra columns to a treatment and a dose column (dose will still contain both start and end dose at this point). Then split the dose column on ' - ' to obtain start_dose and end_dose columns. Drop the intermediate dose column.**

##### Code

In [62]:
# Your cleaning code here
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
treatments_clean = treatments_clean[treatments_clean.dose != "-"]
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str
treatments_clean = treatments_clean.drop('dose', axis=1)

<ipython-input-62-f5d89a8d152b>:5: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str


##### Test

In [63]:
# Your testing code here
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u  
6        37u      42u  
7        31u      38u  
9        30u      36u

#### Adverse reaction should be part of the `treatments` table

##### Define
Merge the *adverse_reaction* column to the `treatments` table, joining on *given name* and *surname*.

*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution.*

##### Code

In [64]:
# Your cleaning code here
treatments_clean = pd.merge(treatments_clean,adverse_reactions_clean,
                            on = ['given_name','surname'],how = 'left')

##### Test

In [65]:
# Your testing code here
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1       skye  gormanston         7.97       7.62          0.35   auralin   
2     sophia      haugen         7.65       7.27          0.38   auralin   
3      eddie      archer         7.89       7.55          0.34   auralin   
4       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end adverse_reaction  
0        41u      48u              NaN  
1        33u      36u              NaN  
2        37u      42u              NaN  
3        31u      38u              NaN  
4        30u      36u              NaN

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*
##### Define
Adverse reactions table is no longer needed so ignore that part. Isolate the patient ID and names in the `patients` table, then convert these names to lower case to join with `treatments`. Then drop the given name and surname columns in the treatments table (so these being lowercase isn't an issue anymore).

In [66]:
##### Code


In [67]:
# cleaning code here
id_names = patients_clean[['patient_id', 'given_name', 'surname']]
id_names.given_name = id_names.given_name.str.lower()
id_names.surname = id_names.surname.str.lower()
treatments_clean = pd.merge(treatments_clean,id_names, on = ['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'],axis=1)

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\generic.py:5303: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


##### Test

In [68]:
# Your testing code here
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0         7.63       7.20          0.43   auralin        41u      48u   
1         7.97       7.62          0.35   auralin        33u      36u   
2         7.65       7.27          0.38   auralin        37u      42u   
3         7.89       7.55          0.34   auralin        31u      38u   
4         7.76       7.37          0.39   auralin        30u      36u   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242  
2              NaN         345  
3              NaN         276  
4              NaN          15

In [69]:
# Patient ID should be the only duplicate column
all_columns = pd.Series(list(patients_clean)+ list(treatments_clean))
all_columns[all_columns.duplicated()]

23    patient_id
dtype: object

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
Convert the zip code column's data type from a float to a string using `astype`, remove the '.0' using string slicing, and pad four digit zip codes with a leading 0.

*Your definition here. Hint: see the "Data Cleaning Process" page.*

##### Code

In [70]:
import numpy as np
# Your cleaning code here
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5,fillchar='0')
# Reconvert NaNs entries that were converted to '0000n' by code above
patients_clean.zip_code = patients_clean.zip_code.replace('0000n',np.nan)

##### Test

In [71]:
# Your testing code here
patients_clean.zip_code.head(20)

0     92390
1     61812
2     68467
3     07095
4     36303
5     32114
6     84728
7     95037
8     64105
9     10011
10    49780
11    37027
12    24504
13    35203
14    74105
15    37660
16    93706
17    30082
18    59324
19    95966
Name: zip_code, dtype: object

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
*Replace height for rows in the `patients` table that have a height of 27 in (there is only one) with 72 in..*

##### Code

In [72]:
# Your cleaning code here
patients_clean.height = patients_clean.height.replace(27, 72)

##### Test

In [73]:
# Your testing code here
patients_clean[patients_clean.height == 27]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, contact, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [74]:
# Confirm the replacement worked
patients_clean[patients_clean.surname == 'Neudorf']

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country                          contact  birthdate  \
4    AL    36303  United States  334-515-7487TimNeudorf@cuvox.de  2/18/1928   

   weight  height   bmi  phone_number                email  
4   192.3      72  26.1  334-515-7487  TimNeudorf@cuvox.de

#### Full state names sometimes, abbreviations other times

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.*

##### Code

In [75]:
# Your cleaning code here

##### Test

In [76]:
# Your testing code here

#### Dsvid Gustafsson

##### Define
*Replace given name for rows in the `patients` table that have a given name of 'Dsvid' with 'David'*

##### Code

In [77]:
# Your cleaning code here
patients_clean.given_name = patients_clean.given_name.replace('Dsvid', 'David')

##### Test

In [78]:
# Your testing code here
patients_clean[patients_clean.surname == 'Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country  \
8  Kansas City    MO    64105  United States   

                                   contact birthdate  weight  height   bmi  \
8  816-265-9578DavidGustafsson@armyspy.com  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
Convert assigned sex and state to categorical data types. Zip code data type was already addressed above. Convert birthdate to datetime data type. Strip the letter 'u' in start dose and end dose and convert those columns to data type integer.

*Your definition here. Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution, [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution, and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution.*

##### Code

In [79]:
#  cleaning code here
# To category
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

# To datetime
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

# Strip u and to integer
treatments_clean.dose_start = treatments_clean.dose_start.str.strip('u').astype(int)
treatments_clean.dose_start = treatments_clean.dose_end.str.strip('u').astype(int)

##### Test

In [80]:
# Your testing code here
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 16 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    503 non-null    int64         
 1   assigned_sex  503 non-null    category      
 2   given_name    503 non-null    object        
 3   surname       503 non-null    object        
 4   address       491 non-null    object        
 5   city          491 non-null    object        
 6   state         491 non-null    object        
 7   zip_code      491 non-null    object        
 8   country       491 non-null    object        
 9   contact       491 non-null    object        
 10  birthdate     503 non-null    datetime64[ns]
 11  weight        503 non-null    float64       
 12  height        503 non-null    int64         
 13  bmi           503 non-null    float64       
 14  phone_number  491 non-null    object        
 15  email         491 non-null    object    

In [90]:
patients_clean.phone_number.head(20)

0          951-719-9170
1     +1 (217) 569-3204
2          402-363-6804
3     +1 (732) 636-8246
4          334-515-7487
5          386-334-5237
6          775-533-5933
7          408 778 3236
8          816-265-9578
9          718 795 9124
10         906-478-8949
11         931 207 0839
12         434-509-2614
13    +1 (205) 417-8095
14         918-712-3469
15         276-225-1955
16         559 765 7836
17         678-829-8578
18         406-775-2696
19         530 532 8397
Name: phone_number, dtype: object

#### Multiple phone number formats

##### Define
Strip all " ", "-", "(", ")", and "+" and store each number without any formatting. Pad the phone number with a 1 if the length of the number is 10 digits (we want country code).

*Your definition here. Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

##### Code

In [91]:
# Your cleaning code here
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')

##### Test

In [92]:
# Your testing code here
patients_clean.phone_number.head()

0    19517199170
1    12175693204
2    14023636804
3    17326368246
4    13345157487
Name: phone_number, dtype: object

#### Default John Doe data

In [93]:
patients_clean[patients_clean.surname == 'Doe']

patient_id assigned_sex given_name surname          address      city  \
215         216         male       John     Doe  123 Main Street  New York   
229         230         male       John     Doe  123 Main Street  New York   
237         238         male       John     Doe  123 Main Street  New York   
244         245         male       John     Doe  123 Main Street  New York   
251         252         male       John     Doe  123 Main Street  New York   
277         278         male       John     Doe  123 Main Street  New York   

    state zip_code        country                      contact  birthdate  \
215    NY    12345  United States  johndoe@email.com1234567890 1975-01-01   
229    NY    12345  United States  johndoe@email.com1234567890 1975-01-01   
237    NY    12345  United States  johndoe@email.com1234567890 1975-01-01   
244    NY    12345  United States  johndoe@email.com1234567890 1975-01-01   
251    NY    12345  United States  johndoe@email.com1234567890 1975-01-01   
277    NY    12345  United States  johndoe@email.com1234567890 1975-01-01   

     weight  height   bmi phone_number              email  
215   180.0      72  24.4  11234567890  johndoe@email.com  
229   180.0      72  24.4  11234567890  johndoe@email.com  
237   180.0      72  24.4  11234567890  johndoe@email.com  
244   180.0      72  24.4  11234567890  johndoe@email.com  
251   180.0      72  24.4  11234567890  johndoe@email.com  
277   180.0      72  24.4  11234567890  johndoe@email.com

##### Define
Remove the non-recoverable John Doe records from the patients table.

*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.*

##### Code

In [94]:
# Your cleaning code here
patients_clean = patients_clean[patients_clean.surname != 'Doe']

##### Test

In [95]:
# Your testing code here
patients_clean.surname.value_counts()

Jakobsen       3
Taylor         3
Cabrera        2
Liễu           2
Grímsdóttir    2
              ..
Glockner       1
Bazarova       1
Mehari         1
Bois           1
Morita         1
Name: surname, Length: 465, dtype: int64

In [98]:
patients_clean[patients_clean.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
242         243         male       John     O'Brian                 NaN   
249         250         male   Benjamin      Mehler                 NaN   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state zip_code        country  \
29   Port Jervis  New York    12771  United States   
219          NaN       NaN      NaN            NaN   
230          NaN       NaN      NaN            NaN   
234          NaN       NaN      NaN            NaN   
242          NaN       NaN      NaN            NaN   
249          NaN       NaN      NaN            NaN   
257          NaN       NaN      NaN            NaN   
264          NaN       NaN      NaN            NaN   
269          NaN       NaN      NaN            NaN   
278          NaN       NaN      NaN            NaN   
282     Rainelle        WV    25962  United States   
286          NaN       NaN      NaN            NaN   
296          NaN       NaN      NaN            NaN   
502         Burr  Nebraska    68324  United States   

                                        contact  birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707 1985-08-01   155.8      67   
219                                         NaN 1978-04-09   237.8      69   
230                                         NaN 1976-09-23   165.9      63   
234                                         NaN 1936-04-07   199.5      65   
242                                         NaN 1957-02-25   205.3      74   
249                                         NaN 1951-10-30   146.5      69   
257                                         NaN 1995-05-17   231.7      69   
264                                         NaN 1989-11-03   158.6      63   
269                                         NaN 1937-10-09   175.2      61   
278                                         NaN 1962-12-16   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com 1960-10-23   206.1      64   
286                                         NaN 1979-04-01   155.3      68   
296                                         NaN 1990-05-14   181.1      63   
502        PatrickGersten@rhyta.com402-848-4923 1954-05-03   138.2      71   

      bmi phone_number                      email  
29   24.4  18458587707  JakobCJakobsen@einrot.com  
219  35.1          NaN                        NaN  
230  29.4          NaN                        NaN  
234  33.2          NaN                        NaN  
242  26.4          NaN                        NaN  
249  21.6          NaN                        NaN  
257  34.2          NaN                        NaN  
264  28.1          NaN                        NaN  
269  33.1          NaN                        NaN  
278  18.4          NaN                        NaN  
282  35.4  13044382648   SandraCTaylor@dayrep.com  
286  23.6          NaN                        NaN  
296  32.1          NaN                        NaN  
502  19.3  14028484923   PatrickGersten@rhyta.com

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
*Remove the Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the `patients` table. These are the nicknames, which happen to also not be in the `treatments` table (removing the wrong name would create a consistency issue between the `patients` and `treatments` table). These are all the second occurrence of the duplicate. These are also the only occurences of non-null duplicate addresses.*

##### Code

In [101]:
# tilde means not: http://pandas.pydata.org/pandas-docs/stable/indexing.html#boolean-indexing
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [102]:
# Your testing code here
patients_clean[patients_clean.address.duplicated()]

patient_id assigned_sex given_name     surname address city state  \
219         220         male         Mỹ       Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth     Knudsen     NaN  NaN   NaN   
234         235       female    Martina   Tománková     NaN  NaN   NaN   
242         243         male       John     O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin      Mehler     NaN  NaN   NaN   
257         258         male        Jin        Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah      Asfour     NaN  NaN   NaN   
269         270       female     Flavia  Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa       Cabán     NaN  NaN   NaN   
286         287         male      Lewis        Webb     NaN  NaN   NaN   
296         297       female        Chỉ         Lâm     NaN  NaN   NaN   

    zip_code country contact  birthdate  weight  height   bmi phone_number  \
219      NaN     NaN     NaN 1978-04-09   237.8      69  35.1          NaN   
230      NaN     NaN     NaN 1976-09-23   165.9      63  29.4          NaN   
234      NaN     NaN     NaN 1936-04-07   199.5      65  33.2          NaN   
242      NaN     NaN     NaN 1957-02-25   205.3      74  26.4          NaN   
249      NaN     NaN     NaN 1951-10-30   146.5      69  21.6          NaN   
257      NaN     NaN     NaN 1995-05-17   231.7      69  34.2          NaN   
264      NaN     NaN     NaN 1989-11-03   158.6      63  28.1          NaN   
269      NaN     NaN     NaN 1937-10-09   175.2      61  33.1          NaN   
278      NaN     NaN     NaN 1962-12-16   124.3      69  18.4          NaN   
286      NaN     NaN     NaN 1979-04-01   155.3      68  23.6          NaN   
296      NaN     NaN     NaN 1990-05-14   181.1      63  32.1          NaN   

    email  
219   NaN  
230   NaN  
234   NaN  
242   NaN  
249   NaN  
257   NaN  
264   NaN  
269   NaN  
278   NaN  
286   NaN  
296   NaN

#### kgs instead of lbs for Zaitseva weight

In [103]:
patients_clean[patients_clean.surname == 'Zaitseva']

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  \
210  Wooster    OH    44691  United States   

                                       contact  birthdate  weight  height  \
210  330-202-2145CamillaZaitseva@superrito.com 1938-11-26    48.8      63   

      bmi phone_number                          email  
210  19.1  13302022145  CamillaZaitseva@superrito.com

##### Define
*Use [advanced indexing](https://stackoverflow.com/a/44913631) to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.*

##### Code

In [88]:
# Your cleaning code here
column_name = 'weight'
mask = patients_clean.surname == 'Zaitseva'
patients_clean.loc[mask,column_name] = weight_kg * 2.20462

##### Test

In [89]:
# Your testing code here